### PR TITLE
MGMT-839 Provide progress reporting of the installed master after piv…

### PR DESCRIPTION
**Pull request by tsorya**
…oting from bootstrap :

Adding assisted-installer-controller and its deployment

### CODE REVIEW

File: src/assisted_installer_controller/assisted_installer_controller.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+func NewController(log *logrus.Logger, cfg ControllerConfig, ops ops.Ops, ic inventory_client.InventoryClient, kc k8s_client.K8SClient) *controller {
+	return &controller{
+		log:              log,
+		ControllerConfig: cfg,
+		ops:              ops,
+		ic:               ic,
+		kc:               kc,
+	}
+}
+
```
Comment (eranco74) [inline src/assisted_installer_controller/assisted_installer_controller.go:46]: Please add doc explaining what it's doing

File: src/assisted_installer_controller/assisted_installer_controller.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+		}
+		c.log.Infof("Found %d nodes: %+v", len(nodes.Items), nodeNameAndCondition)
+		if len(nodes.Items) >= len(assistedInstallerNodes) {
+			c.log.Infof("WaitAndUpdateNodesStatus - Done")
+			cancel()
+		}
+	}, 30*time.Second, apiContext.Done())
+}
+
+func (c *controller) updateHostStatus(hostId string, newStatus string) error {
```
Comment (eranco74) [inline src/assisted_installer_controller/assisted_installer_controller.go:77]: You don't really need this method, you get the same logging in `c.ic.UpdateHostStatus`

File: src/assisted_installer_controller/assisted_installer_controller.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	c.log.Infof("Updating installation status: %s for node: %s", newStatus, hostId)
+	if err := c.ic.UpdateHostStatus(newStatus, hostId); err != nil {
+		c.log.Errorf("Failed to update node installation status, %s", err)
+		return err
+	}
+	return nil
+}
+
+func (c *controller) getNodes() ([]string, *v1.NodeList, error) {
```
Comment (eranco74) [inline src/assisted_installer_controller/assisted_installer_controller.go:86]: I think you can get his once at the start of `WaitAndUpdateNodesStatus` the assistedInstallerNodes are not going to change during the installation.

File: src/assisted_installer_controller/assisted_installer_controller.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+		}
+		for _, node := range nodes.Items {
+			_, ok := nodeNameAndCondition[node.Name]
+			if ok {
+				return
+			}
+			c.log.Infof("Found new joined node %s with id %s", node.Name, node.Status.NodeInfo.SystemUUID)
+			err := c.updateHostStatus(node.Status.NodeInfo.SystemUUID, done)
+			if err == nil {
+				nodeNameAndCondition[node.Name] = node.Status.NodeInfo.SystemUUID //node.Status.Conditions
```
Comment (eranco74) [inline src/assisted_installer_controller/assisted_installer_controller.go:65]: What's this for?

Comment (eranco74) [reply src/assisted_installer_controller/assisted_installer_controller.go:65]: Also should probably rename it to nodeNameToUuidMap.
Why do you need the uuid?

File: src/installer/installer_test.go [hunk 5]
```diff
@@ -104,12 +105,22 @@ var _ = Describe("installer HostRoleMaster role", func() {
...
+		}
 		waitForBootkubeSuccess := func() {
 			mockops.EXPECT().ExecPrivilegeCommand("bash", "-c", "systemctl status bootkube.service | grep 'bootkube.service: Succeeded' | wc -l").Return("1", nil).Times(1)
 		}
 		bootkubeStatusSuccess := func() {
 			mockops.EXPECT().ExecPrivilegeCommand("systemctl", "status", "bootkube.service").Return("1", nil).Times(1)
 		}
+
+		addPolicyForController := func() {
+			fmt.Println("aaaaaaaa")
```
Comment (eranco74) [inline src/installer/installer_test.go:119]: aaaaaaaaa!

File: src/assisted_installer_controller/assisted_installer_controller.go (hunk 3 above)
Comment (tsorya) [reply src/assisted_installer_controller/assisted_installer_controller.go:86]: thought about it, not sure why decided to add it here. will move

File: src/installer/installer.go [hunk 6]
```diff
@@ -193,23 +200,44 @@ func (i *installer) waitForControlPlane(ctx context.Context, kc k8s_client.K8SCl
...
 			i.log.Errorf("Failed to update node installation status, %s", err)
 		}
 	}
 }
 
+func (i *installer) addPolicyForController() error {
+	i.log.Infof("Adding policy for assisted-deployment namespace")
+	err := retry.Do(
+		func() error {
+			_, err := i.ops.ExecPrivilegeCommand("oc", fmt.Sprintf("--kubeconfig=%s", filepath.Join(InstallDir, KubeconfigPath)),
```
Comment (eranco74) [inline src/installer/installer.go:233]: Please use the kube-client

File: src/installer/installer.go [hunk 7]
```diff
@@ -193,23 +200,44 @@ func (i *installer) waitForControlPlane(ctx context.Context, kc k8s_client.K8SCl
...
 			i.log.Errorf("Failed to update node installation status, %s", err)
 		}
 	}
 }
 
+func (i *installer) addPolicyForController() error {
+	i.log.Infof("Adding policy for assisted-deployment namespace")
+	err := retry.Do(
+		func() error {
+			_, err := i.ops.ExecPrivilegeCommand("oc", fmt.Sprintf("--kubeconfig=%s", KubeconfigPath),
```
Comment (tsorya) [inline src/installer/installer.go:233]: i failed to find a way to run this with a client. Will be glad for help. We can open a ticket to move it to client.

File: src/assisted_installer_controller/assisted_installer_controller.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+			}
+			c.log.Infof("Found new joined node %s with id %s, updating its status to %s", node.Name, node.Status.NodeInfo.SystemUUID, done)
+			if err := c.ic.UpdateHostStatus(done, node.Status.NodeInfo.SystemUUID); err != nil {
+				c.log.Errorf("Failed to update node installation status, %s", err)
+				return
+			}
+
+			nodeNameAndUUID[node.Name] = node.Status.NodeInfo.SystemUUID
+		}
+		c.log.Infof("Found %d nodes: %+v", len(nodes.Items), nodeNameAndUUID)
```
Comment (eranco74) [inline src/assisted_installer_controller/assisted_installer_controller.go:75]: Consider logging which nodes are still missing

File: src/assisted_installer_controller/assisted_installer_controller.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+		}
+	}, 30*time.Second, apiContext.Done())
+}
+
+func (c *controller) getInventoryNodes() {
+	c.log.Infof("Getting list of inventory nodes")
+	apiContext, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	wait.Until(func() {
+		var err error
```
Comment (eranco74) [inline src/assisted_installer_controller/assisted_installer_controller.go:88]: Why do you need this?

File: src/assisted_installer_controller/assisted_installer_controller.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+			cancel()
+		}
+	}, 30*time.Second, apiContext.Done())
+}
+
+func (c *controller) getInventoryNodes() {
+	c.log.Infof("Getting list of inventory nodes")
+	apiContext, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	wait.Until(func() {
```
Comment (eranco74) [inline src/assisted_installer_controller/assisted_installer_controller.go:87]: Until what?
you are doing a single API call and return no matter what

File: src/assisted_installer_controller/assisted_installer_controller.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+		if err != nil {
+			c.log.Infof("Failed to get node list from inventory, will retry, %e", err)
+			return
+		}
+		cancel()
+	}, 30*time.Second, apiContext.Done())
+	c.log.Infof("Got list of host from inventory %s", c.assistedInstallerNodes)
+}
+
+func (c *controller) getNodes() (*v1.NodeList, error) {
```
Comment (eranco74) [inline src/assisted_installer_controller/assisted_installer_controller.go:99]: This methid is not required you get the same logging in `c.kc.ListNodes()`

File: src/assisted_installer_controller/assisted_installer_controller.go [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+	nodes, err := c.kc.ListNodes()
+	if err != nil {
+		c.log.Warnf("Still waiting for nodes: %v", err)
```
Comment (eranco74) [inline src/assisted_installer_controller/assisted_installer_controller.go:102]: you are not waiting, and no reason for warning log this is a normal state

File: src/utils/utils.go [hunk 13]
```diff
@@ -14,12 +15,15 @@ import (
 func InitLogger(verbose bool) *logrus.Logger {
 	var log = logrus.New()
 	// log to console and file
-	f, err := os.OpenFile("crawler.log", os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
+	f, err := os.OpenFile("/crawler.log", os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
```
Comment (tsorya) [inline src/utils/utils.go:18]: delete slash, entered by mistake

File: src/utils/utils.go [hunk 14]
```diff
@@ -14,12 +15,15 @@ import (
...
 	if err != nil {
-		log.Fatalf("error opening file: %v", err)
+		log.Fatalf("error aaaaa file: %v", err)
```
Comment (tsorya) [inline src/utils/utils.go:20]: delete it

File: src/installer/installer.go [hunk 15]
```diff
@@ -193,23 +200,44 @@ func (i *installer) waitForControlPlane(ctx context.Context, kc k8s_client.K8SCl
...
 	i.log.Infof("Updating node installation status: %s", newStatus)
 	if i.HostID != "" {
-		if err := i.ic.UpdateHostStatus(newStatus); err != nil {
+		if err := i.ic.UpdateHostStatus(newStatus, i.HostID); err != nil {
 			i.log.Errorf("Failed to update node installation status, %s", err)
 		}
 	}
 }
 
+func (i *installer) addPolicyForController() error {
```
Comment (eranco74) [inline src/installer/installer.go:229]: Please move it to the the kubeclient.go.
and make it generic

File: src/utils/utils.go [hunk 16]
```diff
@@ -14,12 +15,15 @@ import (
 func InitLogger(verbose bool) *logrus.Logger {
```
Comment (tsorya) [inline src/utils/utils.go:15]: delete changes to InitLogger, entered by mistake

File: src/installer/installer.go (hunk 15 above)
Comment (tsorya) [reply src/installer/installer.go:229]: we need it generic? i think we don't. Hope to delete it one day

File: src/assisted_installer_controller/assisted_installer_controller.go (hunk 10 above)
Comment (tsorya) [reply src/assisted_installer_controller/assisted_installer_controller.go:87]: Until cancel, i want to be i have no error and list is full

File: src/assisted_installer_controller/assisted_installer_controller.go (hunk 9 above)
Comment (tsorya) [reply src/assisted_installer_controller/assisted_installer_controller.go:88]: cause i am using c.assistedInstallerNodes and i want to print error too.
i could not to c.assistedInstallerNodes, err := c.ic.GetHostsIds()
thats why i need to create error before

File: src/assisted_installer_controller/assisted_installer_controller.go (hunk 11 above)
Comment (tsorya) [reply src/assisted_installer_controller/assisted_installer_controller.go:99]: right

File: src/installer/installer.go (hunk 15 above)
Comment (eranco74) [reply src/installer/installer.go:229]: Delete it now if you can.
Why not make it generic?
Isn't it just `oc --config... some command`

Comment (tsorya) [reply src/installer/installer.go:229]: it doesn't worth to add dependency on ops to client to run it.

Comment (tsorya) [reply src/installer/installer.go:229]: yap it is. 
Can't delete, still didn't find any other way to do it.
but let it be here. Don't want to add dependency to k8client

File: src/assisted_installer_controller/assisted_installer_controller.go (hunk 10 above)
Comment (tsorya) [reply src/assisted_installer_controller/assisted_installer_controller.go:95]: until there is no error from bm-inventory client

File: src/assisted_installer_controller/assisted_installer_controller.go (hunk 9 above)
Comment (eranco74) [reply src/assisted_installer_controller/assisted_installer_controller.go:96]: Consider returning the inventoryNodes from this method.
You named it `getInventoryNodes` Not `setInventoryNodes`
It should also solve this issue.

File: src/assisted_installer_controller/assisted_installer_controller.go [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+	c.log.Infof("Waiting till all nodes will join and update status to assisted installer")
+
+	c.getInventoryNodes()
+
+	apiContext, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	nodeNameAndUUID := make(map[string]string)
+	wait.Until(func() {
+		nodes, err := c.getNodes()
+		if err != nil {
```
Comment (eranco74) [inline src/assisted_installer_controller/assisted_installer_controller.go:63]: Consider logging the nodes names

Comment (tsorya) [reply src/assisted_installer_controller/assisted_installer_controller.go:63]: There is 2 logs where name will appear. No need for another one

File: src/assisted_installer_controller/assisted_installer_controller.go (hunk 9 above)
Comment (tsorya) [reply src/assisted_installer_controller/assisted_installer_controller.go:96]: explained why

Comment (tsorya) [reply src/assisted_installer_controller/assisted_installer_controller.go:96]: will change name to setInventoryNodes

File: src/main/assisted-installer-controller/assisted_installer_main.go [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+	logger.Infof("Start running assisted-installer with cluster-id %s, host %s , port %d ",
+		Options.ControllerConfig.ClusterID, Options.ControllerConfig.Host, Options.ControllerConfig.Port)
+
+	finisher := assistedinstallercontroller.NewController(logger,
+		Options.ControllerConfig,
+		ops.NewOps(logger),
+		inventory_client.CreateInventoryClient(Options.ControllerConfig.ClusterID, Options.ControllerConfig.Host, Options.ControllerConfig.Port),
+		kc,
+	)
+	finisher.WaitAndUpdateNodesStatus()
```
[the remark this answers is not in the document]
Comment (tsorya) [reply src/main/assisted-installer-controller/assisted_installer_main.go:41]: sure

File: src/installer/installer.go [hunk 19]
```diff
@@ -202,23 +210,44 @@ func (i *installer) waitForControlPlane(ctx context.Context, kc k8s_client.K8SCl
...
 	i.log.Infof("Updating node installation status: %s", newStatus)
 	if i.HostID != "" {
-		if err := i.ic.UpdateHostStatus(newStatus); err != nil {
+		if err := i.ic.UpdateHostStatus(newStatus, i.HostID); err != nil {
 			i.log.Errorf("Failed to update node installation status, %s", err)
 		}
 	}
 }
 
+func (i *installer) addPolicyForController(kc k8s_client.K8SClient) error {
```
Comment (romfreiman) [inline src/installer/installer.go:239]: why this is not a yaml?

**Pull request merged**